File: source/downloads/code/LeetCode/Python/Rotate-List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None
class Solution(object): 
    def rotateRight(self, head, k): 
        if head == None or head.next == None:
            return head
        l = 1 
        i = 0 
        tmp = head 
        # Get the length 
        while tmp.next != None: 
            l += 1 
            tmp = tmp.next 
        dummy_node = ListNode(-1) 
        dummy_node.next = head 
        k = k%l 
        if k == 0: 
            return head 
        while i<l-k: 
            dummy_node = dummy_node.next
            i += 1
        newhead = dummy_node.next 
        dummy_node.next = None 
        # connect the old head to new head 
        tmp = newhead
        while tmp.next != None:
            tmp = tmp.next     
        tmp.next = head
        return newhead

File: downloads/code/LeetCode/Java/Rotate-List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        ListNode prev = new ListNode(-1);
        prev.next = head;
        ListNode cur = prev;
        int i=0;
        int n =0;
        // get n
        while(cur!=null){
            n++;
            cur = cur.next;
        }
        // get new head
        cur = prev;
        while(i<n){
            cur = cur.next;
        }
        prev.next = cur.next;
        cur.next = null;

        return prev.next;
    }
}
